airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/ResourceRequirementsUtils.java

```diff
@@ -0,0 +1,103 @@
+import io.airbyte.config.ActorDefinitionResourceRequirements;
+import io.airbyte.config.JobTypeResourceLimit;
+import io.airbyte.config.JobTypeResourceLimit.JobType;
+import io.airbyte.config.ResourceRequirements;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class ResourceRequirementsUtils {
+
+  /**
+   * Given connection-level resource requirements, actor-definition-level resource requirements,
+   * worker-default resource requirements, and a job type, returns the final resource requirements
+   * generated by merging the provided requirements in hierarchy order.
+   *
+   * Connection-level resource requirements take precendence over actor-definition level resource
+   * requirements. Within the actor-definition level requirements, job-type-specific requirements take
+   * precedence over default definition requirements. Actor-definition level resource requirements
+   * take precedence over worker default resource requirements.
+   *
+   * @param connectionResourceReqs - the resource requirements set on the connection
+   * @param actorDefinitionResourceReqs - the resource requirements set on the actor definition
+   * @param workerDefaultResourceReqs - the default worker resource requirements set in the env
+   *        variables
+   * @param jobType - type of job to extract resource requirements for from the actor definition reqs
+   * @return resource requirements, if present, otherwise an empty ResourceRequirements object.
+   */
+  public static ResourceRequirements getResourceRequirements(@Nullable final ResourceRequirements connectionResourceReqs,
+                                                             @Nullable final ActorDefinitionResourceRequirements actorDefinitionResourceReqs,
+                                                             @Nullable final ResourceRequirements workerDefaultResourceReqs,
+                                                             final JobType jobType) {
+    final ResourceRequirements jobSpecificDefinitionResourceReqs = getResourceRequirementsForJobType(actorDefinitionResourceReqs, jobType)
+        .orElse(null);
+    final ResourceRequirements defaultDefinitionResourceReqs = Optional.ofNullable(actorDefinitionResourceReqs)
+        .map(ActorDefinitionResourceRequirements::getDefault).orElse(null);
+    return mergeResourceRequirements(
+        connectionResourceReqs,
+        jobSpecificDefinitionResourceReqs,
+        defaultDefinitionResourceReqs,
+        workerDefaultResourceReqs);
+  }
+
+  /**
+   * Given a list of resource requirements, merges them together. Earlier reqs override later ones.
+   *
+   * @param resourceReqs - list of resource request to merge
+   * @return merged resource req
+   */
+  public static ResourceRequirements mergeResourceRequirements(final ResourceRequirements... resourceReqs) {
```

Comment:
I wonder if we should keep this particular method private to this class since it'd be so easy to accidentally get the ordering wrong. I think it'd be ideal to keep the ordering logic for the hierarchy strictly bound and enforced by this class.

Outside callers could just pass in `null` to the above public method for any resource requirements they aren't using, right?

Reply:
Yeah good point, I think I agree that it makes sense to have this class enforce a strict ordering so callers can't mess it up. And yeah passing `null` in should work, though I think it would be slightly cleaner to just have another public method that takes in the same arguments but excluding the actor definition ones since thats the only other call we need to support ATM

Reply:
Yeah sure more public methods sounds good to me